rfcs/0001-messages.md

```diff
@@ -0,0 +1,408 @@
+    ..
+}
+```
+
+### Rules
+
+The following rules must be implemented:
+
+#### Header
+
+- `version` must be a supported version.
+- `sender` and `receiver` can't be the same.
+- The `ParticipantId` variants of `sender` and `receiver` must match the message type.
+
+#### Payloads
+
+- Each jubjub type must be validated during deserialization.
+- `share_commitments`:
+  - Length must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+  - Lenght must be at least `MIN_SIGNERS` (`2` signers).
+- `signing_commitments`:
+    - Length must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+    - Signing packages that contain duplicate `ParticipantId`s are invalid. This is implicit in the use of `HashMap`.
+    - Length must be at least `MIN_THRESHOLD` (`2` required signers).
+- `message`: signed messages have a protocol-specific length limit. For Zcash, that limit is the maximum network protocol message length: `2^21` bytes (2 MB).
+
+## Serialization/Deserialization
+
+Each message struct needs to serialize to bytes representation before it is sent through the wire and must deserialize to the same struct (round trip) on the receiver side. We use `serde` and macro derivations (`Serialize` and `Deserialize`) to automatically implement where possible.
+
+This will require deriving serde in several types defined in `frost.rs`. 
+Manual implementation of serialization/deserialization will be located at a new mod `src/frost/serialize.rs`.
+
+### Byte order
+
+Each byte chunk specified below is in little-endian order unless is specified otherwise.
+
+### Header
+
+The `Header` part of the message is 4 bytes total:
+
+Bytes | Field name | Data type
+------|------------|-----------
+1     | msg_type   | u8
+1     | version    | u8
+1     | sender     | u8
+1     | receiver   | u8
+
+### Primitive types
+
+`Payload`s use data types that we need to specify first. We have 2 primitive types inside the payload messages:
+
+#### `Scalar`
+
+`Scalar` is a an alias for `jubjub::Fr`. We use `Scalar::to_bytes` and `Scalar::from_bytes` to get a 32-byte little-endian canonical representation. See https://github.com/zkcrypto/bls12_381/blob/main/src/scalar.rs#L252
+
+#### `AffinePoint`
+
+Much of the math in FROST is done using `jubjub::ExtendedPoint`. But for message exchange `jubjub::AffinePoint`s are a better choice, as their byte representation is smaller.
+
+Conversion from one type to the other is trivial:
+
+https://docs.rs/jubjub/0.6.0/jubjub/struct.AffinePoint.html#impl-From%3CExtendedPoint%3E
+https://docs.rs/jubjub/0.6.0/jubjub/struct.ExtendedPoint.html#impl-From%3CAffinePoint%3E
+
+We use `AffinePoint::to_bytes` and `AffinePoint::from_bytes` to get a 32-byte little-endian canonical representation. See https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L443
+
+Similarly, `VerificationKey`s can be serialized using `<[u8; 32]>::from` and `VerificationKey::from`. See https://github.com/ZcashFoundation/redjubjub/blob/main/src/verification_key.rs#L86
+
```

Comment:
```suggestion
Similarly, `VerificationKey`s can be serialized using `<[u8; 32]>::from` and `VerificationKey::from`. See https://github.com/ZcashFoundation/redjubjub/blob/main/src/verification_key.rs#L86

Multi-byte integers **must not** be used for serialization, because they have different byte orders on different platforms.

```